rasa/shared/core/events.py

```diff
@@ -1548,9 +1546,9 @@ def __repr__(self) -> Text:
             self.action_name, self.policy, self.confidence
         )
 
-    def __str__(self) -> Optional[Text]:
+    def __str__(self) -> Text:
         """Returns event as human readable string."""
-        return self.action_name or self.action_text
+        return self.action_name or self.action_text or ""
```

Comment:
Both `action_name` and `action_text` are of `Optional[Text]` type. alternatively, I can cast both attributes as `str`, and then `None` will appear as string and be acceptable as return type?

Reply:
ok I've tried some time and it doesn't really work. The thing is that I think we need 2 classes (`ActionExecuted` and `ActionEndToEndExecuted`). It would require a larger refactoring.

So my reco here is to cast the return value here to string (not introduce `or ""`), and add something like `if action_name is None and action_text is None: raise ValueError(...)` in the constructor of `ActionExecuted`.

Reply:
(`cast` doesn't really cast the value, it just tells to `mypy` that the value is casted. It doesn't do anything at runtime)

Reply:
Just working on it, had to amend some tests using action events that didn't contain action names after raising `ValueError`.

Reply:
Done now: https://github.com/RasaHQ/rasa/pull/10965/commits/72e76e0449a0bcbf19e2ee802a2f7584f331bf3e